### PR TITLE
fix EHeapElems bug by combining length member with heap array

### LINKED ISSUE #60: EHeapElems is buggy
```
cozy maxbag.ds --java MaxBag.java --timeout=2
```

Then compile and run as https://github.com/CozySynthesizer/cozy/pull/59 indicates

```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3181)
	at java.util.ArrayList.grow(ArrayList.java:265)
	at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:239)
	at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:231)
	at java.util.ArrayList.add(ArrayList.java:462)
	at MaxBag.add(MaxBag.java:183)
	at MaxBagMain.main(MaxBagMain.java:25)
```

I suspect that this is a code-gen bug.

### PR DESCRIPTION
**Pull request by izgzhen**
fix #60

### CODE REVIEW

File: cozy/codegen/cxx.py [hunk 1]
```diff
@@ -274,6 +274,8 @@ def construct_concrete(self, t : Type, e : Exp, out : Exp):
             return SEscape("{indent}{lhs} = {rhs};\n", ["lhs", "rhs"], [out, e])
         elif is_scalar(t):
             return SEscape("{indent}{lhs} = {rhs};\n", ["lhs", "rhs"], [out, e])
+        elif isinstance(t, TTuple):
+            return SEscape("{indent}{lhs} = {rhs};\n", ["lhs", "rhs"], [out, e])
```
Comment (izgzhen) [inline cozy/codegen/cxx.py]: @Calvin-L I believe this also fixes the problem in #70, thus I am a bit confused here

Comment (Calvin-L) [reply cozy/codegen/cxx.py]: See my note in #72 about aliases. I think this solution is dangerous and you really need to create a `new` tuple object.

Comment (izgzhen) [reply cozy/codegen/cxx.py]: > you really need to create a new tuple object.

How exactly? `rhs` will be visited and turned into `new __typeXXX`. However, this is not apparent from the code

Comment (Calvin-L) [reply cozy/codegen/cxx.py]: I believe this has now been correctly addressed in #73. 😄

**Comment by izgzhen** [comment]
@Calvin-L I think I've done working on this -- do you have any feedback?

File: cozy/codegen/cxx.py [hunk 2]
```diff
@@ -374,6 +374,9 @@ def visit_EMakeMap2(self, e):
         self.declare(m, e)
         return m.id
 
+    def visit_EArrayList(self, e):
+        return "(std::vector<{}>())".format(self.visit(e.type.elem_type, ""));
```
Comment (Calvin-L) [inline cozy/codegen/cxx.py]: Semicolon is not needed in Python

File: cozy/codegen/java.py [hunk 3]
```diff
@@ -136,6 +136,9 @@ def body(x):
                 self.end_statement()
         self.end_statement()
 
+    def visit_EArrayList(self, e):
+        return "(new {}[0])".format(self.visit(e.type.elem_type, ""));
```
Comment (Calvin-L) [inline cozy/codegen/java.py]: Semicolon is not needed in Python

**Comment by izgzhen** [comment]
> Additionally, there is one failing test because of an unimplemented case in visit_EArrayIndexOf in cxx.py.

https://github.com/CozySynthesizer/cozy/blob/cb7fd1e78289f6a472bb2ea4647f9ccce2c1fad6/cozy/codegen/cxx.py#L351

It seems like I will need to comment this assertion out -- and even if this is done, generated code will complain about `hash` function: there is no hash function for `std::vector<int>`

**Comment by izgzhen** [comment]
The curious fact is that, `hashCode` can be applied to an integer array in Java, but `std::hash` can't be applied to an integer vector in C++.

**Comment by izgzhen** [comment]
I found that I can just delete the un-compilable gash function definition in C++. it seems not needed.

I think I will first proposed to only generate hash function for struct that actually needs it.

**Comment by izgzhen** [comment]
revised C++ hash struct logic a bit -- only generating Hash function if all subfields are scalar types

**Comment by izgzhen** [comment]
@Calvin-L regarding the ArrayIndexOf, it seems that what you need is to cache evaluated `e.a` inside some variable?

**Comment by Calvin-L** [comment]
That's right. I put that `assert` in place to make sure that someday we store the result of `e.a` in a variable to avoid recomputing it over and over.

**Comment by izgzhen** [comment]
> That's right. I put that assert in place to make sure that someday we store the result of e.a in a variable to avoid recomputing it over and over.

I figured that caching inside a variable is a bit overkill in this case - I found that in C++ assignment might copy the vector. In my case, the `e.a` is just a `ETupleGet(EVar(...), 1)`. If we assign this expression to a new variable, I am afraid that there will be extra copying. So I just add my case as another special case which doesn't incur repeated computation, and also avoiding overhead of copying.

**Comment by izgzhen** [comment]
a lot of tests are broken by my hash implementation -- @Calvin-L maybe we can do this in a separate commit later?

**Comment by izgzhen** [comment]
@Calvin-L ping

File: cozy/structures/arrays.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 from cozy.syntax import Type, Exp, Stm
 
 TArray = declare_case(Type, "TArray", ["elem_type"])
+EArrayList = declare_case(Exp, "EArrayList", [])
```
[the remark this answers is not in the document]
Comment (Calvin-L) [reply cozy/structures/arrays.py]: It isn't documented, and it probably isn't even implemented correctly, but my intention had been that `SArrayAlloc` declares a new array---its `.a` field must be a variable and the code generator must declare it like `self.declare` would. `SArrayRealloc` is used to change the size of an already-allocated array. There is no construct to replace the contents of an array all at once.

[the remark this answers is not in the document]
Comment (izgzhen) [reply cozy/structures/arrays.py]: yeah, I see your intention. However, currently this API assume that `a` is already declared https://github.com/CozySynthesizer/cozy/blob/master/cozy/codegen/cxx.py#L362

**Comment by izgzhen** [comment]
@Calvin-L I intend to revert my attempt to implement hash function for array, and leave it as a future work since it breaks too many regression test -- what do you think?

**Comment by Calvin-L** [comment]
I would much prefer not to push code that breaks tests. If you have a follow-up in mind that fixes them, just add it to this PR.

Also, a quick note:

Many of the tests are failing on this PR with messages like `At (x == 0): cannot unify types Int and std::size_t (due to comparison)`. I suspect this is because Cozy is somehow changing the type of a global constant somewhere. Off the cuff, I'd say it isn't anything you committed, but rather [this line in `cxx.py`](https://github.com/CozySynthesizer/cozy/blob/9d6582f3ae647b6a6e16e963218dc5120d2191fb/cozy/codegen/cxx.py#L821). Seems that past-me did not follow his own advice about creating shallow copies when adjusting the type of an expression that might alias with others. 🤣 Would you mind adjusting that line to `self.declare(shallow_copy(s).var.with_type(t), s.val)` and see if more tests pass?

File: cozy/structures/arrays.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Calvin-L) [reply cozy/structures/arrays.py]: Great. Then for now, your change seems good.

**Comment by Calvin-L** [comment]
(FWIW: I am eager to get this merged, I'd just like to do it with a passing test suite. There is good stuff in this PR, and the changes I am putting together for #49 depend on some of the changes here.)

**Comment by mernst** [comment]
I agree that we should never push code to master that breaks tests.

**Comment by izgzhen** [comment]
> Would you mind adjusting that line to self.declare(shallow_copy(s.var).with_type(t), s.val) and see if more tests pass?

Pushed update and it seems fixing it (also, with just the fix you suggested, `solver.py` can't deal with `TNative(std::size_t)`, so I rewrite some types in hashing function from `TNative(std::size_t)` to `INT`. It indeed helps.

Thanks a lot!

**Comment by izgzhen** [comment]
@Calvin-L tests are passed!

**Pull request merged**